### PR TITLE
Fixes for #6231 & #6315

### PR DESCRIPTION
**Pull request by phantomjinx**
Done a lot of testing with different permutations, including
1. go does not exist
2. go is too low a version
3. Neither docker nor go installed
4. go is a higher version (1.13)
5. docker does not exist
6. docker not running - docker info fails
7. operator-sdk not installed
8. GO_PATH not set
9. GO_PATH is wrong
10. Project not under GO_PATH

All or some of these will basically warn/stop the operator or camel-k from building.

### CODE REVIEW

**Comment by phantomjinx** [comment]
Added in the easiest workaround at the moment for #6315. If we want to explore a better solution then no problem.

File: install/operator/.lib.sh [hunk 1]
```diff
@@ -39,71 +39,51 @@ readopt() {
...
-                echo "          please move this project under the \$GOPATH so that :'$(pwd)'"
-                echo "          is located at '$GOPATH/src/${gopackage}'"
-                echo
-            else
-                operator-sdk generate k8s
-                # operator-sdk generate openapi
-            fi
+
+        local hassdk=$(operatorsdk_is_available)
+        if [ "$hassdk" = "OK" ]; then
```
Comment (zregvart) [inline install/operator/.lib.sh]:
```suggestion
        if [ "$hassdk" == "OK" ]; then
```

File: install/operator/.lib.sh [hunk 2]
```diff
@@ -39,71 +39,51 @@ readopt() {
...
-            if [[ "$(which go)" == "" ]] ; then
-                echo 'error: you must have either docker or go installed to build this project.'
-                exit 1
-            else
-                strategy="go"
-            fi
+    local hasgo=$(go_is_available)
+    local hasdocker=$(docker_is_available)
+
+    if [ "$strategy" = "auto" ] ; then
```
Comment (zregvart) [inline install/operator/.lib.sh]:
```suggestion
    if [ "$strategy" == "auto" ] ; then
```

File: install/operator/.lib.sh [hunk 3]
```diff
@@ -39,71 +39,51 @@ readopt() {
...
+        if [ "$hasgo" = "OK" ]; then
```
Comment (zregvart) [inline install/operator/.lib.sh]:
```suggestion
        if [ "$hasgo" == "OK" ]; then
```

File: install/operator/.lib.sh [hunk 4]
```diff
@@ -39,71 +39,51 @@ readopt() {
...
+            # Prefer go over docker - TODO: do we prefer this?
+            strategy="go"
+        elif [ "$hasdocker" = "OK" ]; then
```
Comment (zregvart) [inline install/operator/.lib.sh]:
```suggestion
        elif [ "$hasdocker" == "OK" ]; then
```

File: install/operator/.lib.sh [hunk 5]
```diff
@@ -177,15 +165,31 @@ build_image()
...
 
+    local hasdocker=$(docker_is_available)
+    local hasoc=$(is_oc_available)
+
     if [ "$strategy" == "auto" ] ; then
-        strategy=docker
-        if [ -n "$(which oc)" ] ; then
-            strategy=s2i
+
+        if [ "$hasoc" = "OK" -o "$hasoc" = "$SETUP_MINISHIFT" ]; then
```
Comment (zregvart) [inline install/operator/.lib.sh]:
```suggestion
        if [ "$hasoc" == "OK" -o "$hasoc" == "$SETUP_MINISHIFT" ]; then
```

File: install/operator/.lib.sh [hunk 6]
```diff
@@ -177,15 +165,31 @@ build_image()
...
+            strategy="s2i"
+        elif [ "$hasdocker" = "OK" ]; then
```
Comment (zregvart) [inline install/operator/.lib.sh]:
```suggestion
        elif [ "$hasdocker" == "OK" ]; then
```

File: install/operator/.lib.sh [hunk 7]
```diff
@@ -177,15 +165,31 @@ build_image()
...
+            strategy="docker"
+        else
+            echo "ERROR: Building an image requires either 'docker' or 'oc' commands to be installed."
+            exit 1
         fi
     fi
 
     case "$strategy" in
     "s2i")
+        if [ "$hasoc" = "$SETUP_MINISHIFT" ]; then
```
Comment (zregvart) [inline install/operator/.lib.sh]:
```suggestion
        if [ "$hasoc" == "$SETUP_MINISHIFT" ]; then
```

**Comment by phantomjinx** [comment]
> Couple of typos found. One concept we have used thus far in the build scripts was to call functions via `check_error`, this departs from this a bit. Would it be possible to replace (some of?) these comparisons of `[ "$..." != "OK" ]`/`[ "$..." == "OK" ]` with `check_error $...`?

check_error calls `exit 1` if thrre is an ERROR. In these circumstances, this is not appropriate as a failure can allow another logic route to be chosen with `else` block.

File: install/operator/.lib.sh (hunk 1 above)
Comment (phantomjinx) [reply install/operator/.lib.sh]: Didn't the scripts need to be shell agnostic and avoid specific implementations?

Comment (zregvart) [reply install/operator/.lib.sh]: Even if that was the case, which I'm fairly certain it isn't, that ship has sailed, we're very bash-specific in our shell scripts.

Comment (phantomjinx) [reply install/operator/.lib.sh]: Was just something Roland commented on in one of my first PRs late last year.

Comment (zregvart) [reply install/operator/.lib.sh]: I'm sure he would want us to be better at that, I'm not sure if that's even possible, I think we're using quite a lot of bashisms at this point; also not sure if that's required bash is pretty much universal `sh`, the only use case would be within a busybox based distribution, and I don't think we have that use case.

**Comment by phantomjinx** [comment]
@zregvart 
Are you good with the explanation above or do you want to discuss further?

**Pull request merged**